src/main/kotlin/org/rust/lang/core/psi/RsPsiTreeChangeListener.kt

```diff
@@ -0,0 +1,225 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.psi
+
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiTreeChangeEvent
+import com.intellij.psi.PsiTreeChangeListener
+import com.intellij.psi.impl.PsiTreeChangeEventImpl
+import java.util.*
+
+sealed class RsPsiTreeChangeEvent {
+    /**
+     * Event may be related to changes in a file system, e.g. file create/delete/movement.
+     * In this case the property is set to null.
+     */
+    open val file: PsiFile? get() = null
+
+    sealed class ChildAdd(
```

Comment:
Maybe `ChildAdded` name is better?

Reply:
`ChildAdded.Before` looks weird, I guess.

Reply:
For me, `ChildAdded.Before` and `ChildAdd.Before` are at the same level of weirdness
Nevertheless, I read `ChildAdd` as a command to add a child while `ChildAdded` looks like as some event on child addition. That's why I'd prefer the second variant